lib/logger/lib/logger.ex

```diff
@@ -459,41 +459,44 @@ defmodule Logger do
   """
   @spec metadata(metadata) :: :ok
   def metadata(keyword) do
-    {enabled?, metadata} = __metadata__()
-    Process.put(@metadata, {enabled?, into_metadata(keyword, metadata)})
-    :ok
-  end
+    case :logger.get_process_metadata() do
+      :undefined ->
+        :logger.set_process_metadata(filter_out_nils(keyword))
 
-  defp into_metadata([], metadata), do: metadata
-  defp into_metadata(keyword, metadata), do: into_metadata(keyword, [], metadata)
+      map when is_map(map) ->
+        merged = Map.merge(map, Map.new(keyword))
+        metadata = filter_out_nils(merged)
 
-  defp into_metadata([{key, nil} | keyword], prepend, metadata) do
-    into_metadata(keyword, prepend, :lists.keydelete(key, 1, metadata))
-  end
+        :logger.set_process_metadata(metadata)
+    end
 
-  defp into_metadata([{key, _} = pair | keyword], prepend, metadata) do
-    into_metadata(keyword, [pair | prepend], :lists.keydelete(key, 1, metadata))
+    :ok
   end
 
-  defp into_metadata([], prepend, metadata) do
-    prepend ++ metadata
+  defp filter_out_nils(enumerable) do
+    for {_k, v} = elem <- enumerable, v != nil, into: %{}, do: elem
   end
 
   @doc """
   Reads the current process metadata.
   """
   @spec metadata() :: metadata
   def metadata() do
-    __metadata__() |> elem(1)
+    case :logger.get_process_metadata() do
+      :undefined -> []
+      map when is_map(map) ->
+        map
+        |> Map.drop([:pid, :gl, :time, :mfa, :file, :line, :domain, :report_cb])
```

Comment:
Are you sure those are returned from the process metadata?

Reply:
No, this probably should not filter them out.